src/utils/leak.js

```diff
@@ -586,7 +586,337 @@ export class LeakImplementation {
+        computeFromValue(fn.$BoundTargetFunction);
+        computeFromValue(fn.$BoundThis);
+        for (let boundArg of fn.$BoundArguments) computeFromValue(boundArg);
+        return;
+      }
+
+      invariant(
+        !(fn instanceof NativeFunctionValue),
+        "all native function values should have already been created outside this pure function"
+      );
+
+      // TODO: Add items to nonLocalReadBindings in passing
+      let nonLocalReadBindings = nonLocalReadBindingsOfFunction(fn);
+      computeFromBindings(fn, nonLocalReadBindings);
+    }
+
+    function computeFromObjectPropertiesWithComputedNames(absVal: AbstractValue): void {
+      if (absVal.kind === "widened property") return;
+      if (absVal.kind === "template for prototype member expression") return;
+      notSupportedForTransitiveMaterialization();
+    }
+
+    function computeFromMap(val: ObjectValue): void {
+      let kind = val.getKind();
+
+      let entries;
+      if (kind === "Map") {
+        entries = val.$MapData;
+      } else {
+        invariant(kind === "WeakMap");
+        entries = val.$WeakMapData;
+      }
+      invariant(entries !== undefined);
+      let len = entries.length;
+
+      for (let i = 0; i < len; i++) {
+        let entry = entries[i];
+        let key = entry.$Key;
+        let value = entry.$Value;
+        if (key === undefined || value === undefined) continue;
+        computeFromValue(key);
+        computeFromValue(value);
+      }
+    }
+
+    function computeFromSet(val: ObjectValue): void {
+      let kind = val.getKind();
+
+      let entries;
+      if (kind === "Set") {
+        entries = val.$SetData;
+      } else {
+        invariant(kind === "WeakSet");
+        entries = val.$WeakSetData;
+      }
+      invariant(entries !== undefined);
+      let len = entries.length;
+
+      for (let i = 0; i < len; i++) {
+        let entry = entries[i];
+        if (entry === undefined) continue;
+
+        computeFromValue(entry);
+      }
+    }
+
+    function nonLocalReadBindingsOfFunction(func: FunctionValue) {
+      // unboundWrites is currently not used, but we leave it in place
+      // to reuse the function closure visitor implemented for leaking
+      // Notice the invariant that ensures that it is empty
+      let functionInfo = {
+        unboundReads: new Set(),
+        unboundWrites: new Set(),
+      };
+
+      invariant(func instanceof ECMAScriptSourceFunctionValue);
+
+      let formalParameters = func.$FormalParameters;
+      invariant(formalParameters != null);
+
+      let code = func.$ECMAScriptCode;
+      invariant(code != null);
+
+      traverse(
+        t.file(t.program([t.expressionStatement(t.functionExpression(null, formalParameters, code))])),
+        LeakedClosureRefVisitor,
+        null,
+        functionInfo
+      );
+      traverse.cache.clear();
+
+      // If the function is not pure, we should not be here
+      invariant(functionInfo.unboundWrites.size === 0);
```

Comment:
This is not true. There an assumption based on a _dynamic_ analysis that this should not happen, but here we are doing a _static_ traversal where we might indeed find unbound writes.